corehq/apps/sms/api.py

```diff
@@ -220,35 +259,47 @@ def process_sms_registration(msg):
+    To have a mobile worker register itself, do the following:
 
         1) Select "Enable Mobile Worker Registration via SMS" in project settings.
 
-        2) Text in "join <domain> worker", where <domain> is the domain to join. If the
-        sending number does not exist in the system, a PendingCommCareUser object will be
-        created, tied to that number.
+        2) Text in "join <domain> worker <username>", where <domain> is the domain to join and <username> is the
+        requested username.  If the username doesn't exist it will be created, otherwise the registration will error.
+        If the username argument is not specified, the username will be the mobile number
+
         The "join" and "worker" keywords can be any keyword in REGISTRATION_KEYWORDS and
-        REGISTRATION_MOBILE_WORKER_KEYWORDS, respectively. This is meant to support multiple 
+        REGISTRATION_MOBILE_WORKER_KEYWORDS, respectively. This is meant to support multiple
         translations.
-
-        3) A domain admin will have to approve the addition of the mobile worker before
-        a CommCareUser can actually be created.
     """
     registration_processed = False
     text_words = msg.text.upper().split()
     keyword1 = text_words[0] if len(text_words) > 0 else ""
     keyword2 = text_words[1].lower() if len(text_words) > 1 else ""
     keyword3 = text_words[2] if len(text_words) > 2 else ""
+    keyword4 = text_words[3] if len(text_words) > 3 else ""
+    cleaned_phone_number = strip_plus(msg.phone_number)
     if keyword1 in REGISTRATION_KEYWORDS and keyword2 != "":
         domain = Domain.get_by_name(keyword2, strict=True)
         if domain is not None:
             if domain_has_privilege(domain, privileges.INBOUND_SMS):
                 if keyword3 in REGISTRATION_MOBILE_WORKER_KEYWORDS and domain.sms_mobile_worker_registration_enabled:
-                    #TODO: Register a PendingMobileWorker object that must be approved by a domain admin
-                    pass
+                    if keyword4 != '':
+                        username = keyword4
+                    else:
+                        username = cleaned_phone_number
+                    try:
+                        username = process_username(username, domain)
+                        password = random_password()
+                        new_user = CommCareUser.create(domain.name, username, password)
+                        new_user.add_phone_number(cleaned_phone_number)
+                        new_user.save_verified_number(domain.name, cleaned_phone_number, True, None)
+                        new_user.save()
+                        registration_processed = True
+                    except ValidationError:
+                        send_sms(domain.name, None, cleaned_phone_number, 'This username is already taken, please '
+                                                                          'select another')
```

Comment:
We should respond with the actual error - looks like there could be more errors than just a duplicate username. Also, it would be better if all the response error messages were defined in https://github.com/dimagi/commcare-hq/blob/register-mobile-worker-sms/corehq/apps/sms/messages.py. Then we can call the get_message function (https://github.com/dimagi/commcare-hq/blob/register-mobile-worker-sms/corehq/apps/sms/messages.py#L51) to retrieve the actual text to send.